### PR TITLE
Support ignoring shebangs

### LINKED ISSUE #116: comments rule with require-starting-space: true should special case shebang
Some yaml files might start with a shebang on their first line, e.g. ansible playbooks with `#!/usr/bin/env ansible-playbook`. The `require-starting-space` option currently complains about such lines, asking you to break them by inserting a space between `#` and `!`, it could special case them.

### PR DESCRIPTION
**Pull request by mattiasb**
Some usages of YAML (like Ansible) supports running the file as a script.

Support (by default) an ignore-shebangs setting for the comments module.

Fixes #116 - comments rule with require-starting-space: true should special case shebang

### CODE REVIEW

File: yamllint/rules/comments.py [hunk 1]
```diff
@@ -67,6 +70,7 @@
 ID = 'comments'
 TYPE = 'comment'
 CONF = {'require-starting-space': bool,
+        'ignore-shebang': bool,
```
Comment (adrienverge) [inline yamllint/rules/comments.py]: `ignore-shebangs` (with an **s** at the end) would be more consistent with other rules, in my opinion. What do you think?

File: yamllint/rules/comments.py [hunk 2]
```diff
@@ -21,6 +21,9 @@
 
 * Use ``require-starting-space`` to require a space character right after the
   ``#``. Set to ``true`` to enable, ``false`` to disable.
+* Use ``ignore-shebang`` to ignore a shebang at the beginning of the file.
+  (https://en.wikipedia.org/wiki/Shebang_(Unix)) when checking
+  ``require-starting-space``
```
Comment (adrienverge) [inline yamllint/rules/comments.py]: There's an extra `.` on line 24, and a `.` is missing on line 26. Also, are you sure the URL is going to be correctly rendered in RST? (You can use `python3 setup.py build_sphinx` to generate HTML docs.)

File: yamllint/rules/comments.py (hunk 1 above)
Comment (mattiasb) [reply yamllint/rules/comments.py]: I agree. :)

File: yamllint/rules/comments.py (hunk 2 above)
Comment (mattiasb) [reply yamllint/rules/comments.py]: No idea about RST :) I'll read up on it!

**Comment by mattiasb** [comment]
Thanks for the review! I'll get back on this around monday if I don't find some time earlier. :)

File: yamllint/conf/default.yaml [hunk 3]
```diff
@@ -21,6 +21,7 @@ rules:
   comments:
     level: warning
     require-starting-space: true
+    ignore-shebang: true
```
[the remark this answers is not in the document]
Comment (adamchainz) [reply yamllint/conf/default.yaml]: In discussion in #116 earlier this year, we concluded the opposite :)

[the remark this answers is not in the document]
Comment (adrienverge) [reply yamllint/conf/default.yaml]: Oh! Thanks for reminding this discussion @adamchainz, it was completely out of my head 🙂

We didn't conclude anything, but you're right: it could be on. I guess no one would blame us for that.

[the remark this answers is not in the document]
Comment (mattiasb) [reply yamllint/conf/default.yaml]: I'll leave this on then unless I hear anything. :)

**Comment by mattiasb** [comment]
Fixed the review comments. No idea on how to fix the backwards compatibility issue you mentioned. I have a limited amount of time I can spend on this but if it's not super invasive I can try to give it a go some time. Could you explain what I'd need to do to get this fixed?

**Comment by mattiasb** [comment]
Sorry for bugging you, but any hints? :)

**Comment by adrienverge** [comment]
Hi @mattiasb,

I'm very sorry about the non-response since last month: I really lacked time.

I've worked on the backward-compatibility issue (see https://github.com/adrienverge/yamllint/pull/157), it should do the job. I still need to test that it doesn't break anything on large projects (Ansible, OpenStack...)

Once merged, a slight rework will be needed on your side (to put the new option default value inside `rules/comments.py` instead of `conf/default.yaml`).

**Comment by mattiasb** [comment]
Cool! 

I'll wait for that to be merged and will get back to this then! :)

File: yamllint/rules/comments.py [hunk 4]
```diff
@@ -21,6 +21,9 @@
 
 * Use ``require-starting-space`` to require a space character right after the
   ``#``. Set to ``true`` to enable, ``false`` to disable.
+* Use ``ignore-shebangs`` to ignore a
+  `shebang <https://en.wikipedia.org/wiki/Shebang_(Unix>`_ at the beginning of
```
Comment (adrienverge) [inline yamllint/rules/comments.py]: The link seems to be missing a final `)`: it leads to a "missing article" page.

**Comment by adrienverge** [comment]
#157 merged, feel free to update when you want :)

**Comment by mattiasb** [comment]
There! :)

**Comment by adrienverge** [comment]
Looks great!

I allowed myself to push to your branch to add a few edits. If you're OK with these changes, I'll merge and create a new release.

```diff
comments: Fix ignore-shebangs option on corner cases

diff --git a/tests/rules/test_comments.py b/tests/rules/test_comments.py
index 742650b..ad18a0e 100644
--- a/tests/rules/test_comments.py
+++ b/tests/rules/test_comments.py
@@ -84,28 +84,43 @@ class CommentsTestCase(RuleTestCase):
         conf = ('comments:\n'
                 '  require-starting-space: true\n'
                 '  ignore-shebangs: false\n'
-                'comments-indentation: disable\n')
+                'comments-indentation: disable\n'
+                'document-start: disable\n')
         self.check('#!/bin/env my-interpreter\n',
                    conf, problem1=(1, 2))
+        self.check('# comment\n'
+                   '#!/bin/env my-interpreter\n', conf,
+                   problem1=(2, 2))
         self.check('#!/bin/env my-interpreter\n'
                    '---\n'
                    '#comment\n'
                    '#!/bin/env my-interpreter\n'
                    '', conf,
                    problem1=(1, 2), problem2=(3, 2), problem3=(4, 2))
+        self.check('#! not a shebang\n',
+                   conf, problem1=(1, 2))
+        self.check('key:  #!/not/a/shebang\n',
+                   conf, problem1=(1, 8))
 
     def test_ignore_shebang(self):
         conf = ('comments:\n'
                 '  require-starting-space: true\n'
                 '  ignore-shebangs: true\n'
-                'comments-indentation: disable\n')
+                'comments-indentation: disable\n'
+                'document-start: disable\n')
         self.check('#!/bin/env my-interpreter\n', conf)
+        self.check('# comment\n'
+                   '#!/bin/env my-interpreter\n', conf,
+                   problem1=(2, 2))
         self.check('#!/bin/env my-interpreter\n'
                    '---\n'
                    '#comment\n'
-                   '#!/bin/env my-interpreter\n'
-                   '', conf,
+                   '#!/bin/env my-interpreter\n', conf,
                    problem2=(3, 2), problem3=(4, 2))
+        self.check('#! not a shebang\n',
+                   conf, problem1=(1, 2))
+        self.check('key:  #!/not/a/shebang\n',
+                   conf, problem1=(1, 8))
 
     def test_spaces_from_content(self):
         conf = ('comments:\n'
diff --git a/yamllint/rules/comments.py b/yamllint/rules/comments.py
index 4777d02..0122838 100644
--- a/yamllint/rules/comments.py
+++ b/yamllint/rules/comments.py
@@ -64,6 +64,8 @@ Use this rule to control the position and formatting of comments.
 """
 
 
+import re
+
 from yamllint.linter import LintProblem
 
 
@@ -92,7 +94,8 @@ def check(conf, comment):
         if text_start < len(comment.buffer):
             if (conf['ignore-shebangs'] and
                     comment.line_no == 1 and
-                    comment.buffer[text_start] == '!'):
+                    comment.column_no == 1 and
+                    re.match(r'^!\S', comment.buffer[text_start:])):
                 return
             elif comment.buffer[text_start] not in (' ', '\n', '\0'):
                 column = comment.column_no + text_start - comment.pointer
```

**Comment by mattiasb** [comment]
Looks good! :)

**Comment by adamchainz** [comment]
Yay! Thanks @mattiasb and @adrienverge

**Comment by mattiasb** [comment]
Woop!
